internal/bridge/crunchybridgecluster/delete_test.go

```diff
@@ -0,0 +1,140 @@
+		cluster := testCluster()
+		cluster.Namespace = ns
+		cluster.Status.ID = "1234"
+		cluster.Spec.ClusterName = "bridge-cluster-1"
+		assert.NilError(t, tClient.Create(ctx, cluster))
+
+		// Run handleDelete
+		controllerResult, err := reconciler.handleDelete(ctx, cluster, "9012")
+		assert.NilError(t, err)
+		assert.Check(t, controllerResult == nil)
+
+		// Make sure that finalizer was added
+		assert.Check(t, controllerutil.ContainsFinalizer(cluster, finalizer))
+
+		// Send delete request to kubernetes
+		assert.NilError(t, tClient.Delete(ctx, cluster))
+
+		// Get cluster from kubernetes and assert that the deletion timestamp was added
+		assert.NilError(t, tClient.Get(ctx, client.ObjectKeyFromObject(cluster), cluster))
+		assert.Check(t, !cluster.ObjectMeta.DeletionTimestamp.IsZero())
+
+		// Run handleDelete again to delete from Bridge
+		cluster.Status.ID = "1234"
+		controllerResult, err = reconciler.handleDelete(ctx, cluster, "9012")
+		assert.NilError(t, err)
+		assert.Equal(t, *controllerResult, ctrl.Result{RequeueAfter: 1 * time.Second})
+		assert.Equal(t, len(testBridgeClient.Clusters), 1)
+		assert.Equal(t, testBridgeClient.Clusters[0].ClusterName, "bridge-cluster-2")
+
+		// Run handleDelete one last time to remove finalizer
+		controllerResult, err = reconciler.handleDelete(ctx, cluster, "9012")
+		assert.NilError(t, err)
+		assert.Equal(t, *controllerResult, ctrl.Result{})
+
+		// Make sure that finalizer was removed
+		assert.Check(t, !controllerutil.ContainsFinalizer(cluster, finalizer))
+	})
+
+	t.Run("UnsuccessfulDeletion", func(t *testing.T) {
+		cluster := testCluster()
+		cluster.Namespace = ns
+		cluster.Status.ID = "2345"
+		cluster.Spec.ClusterName = "bridge-cluster-2"
+		assert.NilError(t, tClient.Create(ctx, cluster))
+
+		// Run handleDelete
+		controllerResult, err := reconciler.handleDelete(ctx, cluster, "9012")
+		assert.NilError(t, err)
+		assert.Check(t, controllerResult == nil)
+
+		// Make sure that finalizer was added
+		assert.Check(t, controllerutil.ContainsFinalizer(cluster, finalizer))
+
+		// Send delete request to kubernetes
+		assert.NilError(t, tClient.Delete(ctx, cluster))
+
+		// Get cluster from kubernetes and assert that the deletion timestamp was added
+		assert.NilError(t, tClient.Get(ctx, client.ObjectKeyFromObject(cluster), cluster))
+		assert.Check(t, !cluster.ObjectMeta.DeletionTimestamp.IsZero())
+
+		// Run handleDelete again to attempt to delete from Bridge, but provide bad api key
+		cluster.Status.ID = "2345"
+		controllerResult, err = reconciler.handleDelete(ctx, cluster, "bad_api_key")
+		assert.ErrorContains(t, err, "boom")
+		assert.Equal(t, *controllerResult, ctrl.Result{})
+
+		// Run handleDelete a couple times with good api key so test can cleanup properly
+		// delete from bridge
+		_, err = reconciler.handleDelete(ctx, cluster, "9012")
+		assert.NilError(t, err)
+
+		// remove finalizer
+		_, err = reconciler.handleDelete(ctx, cluster, "9012")
+		assert.NilError(t, err)
```

Comment:
Why are we running this twice without any changes in between? Do we expect that multiple runs of `handleDelete` will do different things? 

I see that in the successful delete we do the same thing.

Reply:
Yeah if you look at the `handleDelete` function, when a `DeletionTimestamp` is applied to the CR, the first time `handleDelete` is run it sends the request to Bridge to delete the cluster. The response *should* have `deletedAlready` as `false` so it requeues. When `handleDelete` gets run on the requeue loop, the request is made to Bridge again, and the response *should* now result in `deletedAlready` being `true`, so we then remove the finalizer... 

TL;DR: we don't want to remove the finalizer until we're sure that the cluster has been deleted from Bridge, so we have to do multiple calls/reconcile loops.

Reply:
Ah, that makes sense. I think that TL;DR as a comment would be helpful